### PR TITLE
Batch Kafka consumer

### PR DESCRIPTION
**Pull request by z0isch**
Add ability to handle batches for our kafka consumers

### CODE REVIEW

File: library/Freckle/App/Kafka/Consumer.hs [hunk 1]
```diff
@@ -164,31 +164,39 @@ runConsumer
...
+            logExMay
+              "Unable to store offset"
+              (silenceNoOffsetError <$> storeOffsetMessage consumer r)
+
+      inSpan "kafka.consumer.message.commit" defaultSpanArguments $
+        -- Do a best-effort commit to the broker. If this fails and
+        -- we crash or shutdown, the commit in the onFinish handler
+        -- will pick it up.
+        void $
+          commitAllOffsets OffsetCommitAsync consumer
```
Comment (pbrisbin) [inline library/Freckle/App/Kafka/Consumer.hs:196]: I believe the `_NO_OFFSETS` error comes from here (when you commit with nothing stored), so you'd want to move your silence function, right?

File: library/Freckle/App/Kafka/Consumer.hs [hunk 2]
```diff
@@ -164,31 +164,39 @@ runConsumer
      , HasCallStack
      )
   => Timeout
+  -> Int
   -> (a -> m ())
```
Comment (z0isch) [inline library/Freckle/App/Kafka/Consumer.hs:168]: Is this the right shape for the handler? Should it be `[a] -> m ()` instead? 

Seems like it'd more flexible for the client to do "smart" things if we hand them the entire batch. I'm thinking things like doing concurrent things or grouping stuff together.

**Review by pbrisbin — APPROVED** [review]
So this re-introduces some duplicate processing. If we're deploying a second consumer can come up and would reprocess any messages the existing consumers have stored but not yet committed, which is up to `batchSize` messages.

Seems like an OK middle-ground with out it was before. However, I still feel like we could've arrived at the same rough behavior in this regard if we had simply reduced the auto-commit setting to do so more frequently, no?

File: library/Freckle/App/Kafka/Consumer.hs (hunk 1 above)
Comment (z0isch) [reply library/Freckle/App/Kafka/Consumer.hs:196]: I'm seeing the `Unable to commit offsets` message which I think comes from the `logExMay "Unable to store offset"` when we `storeOffsetMessage`

**Comment by pbrisbin** [comment]
If we wanted to keep the reduction in duplicate processing, I think still doing a store+commit on every message, but now with batching could also help things... since you're polling less frequently it would take pressure off the broken even though you're committing more frequently. WDYT?

File: library/Freckle/App/Kafka/Consumer.hs (hunk 1 above)
Comment (pbrisbin) [reply library/Freckle/App/Kafka/Consumer.hs:196]: Sorry, I'm confused... you're saying you see `"Unable to commit offsets"` but you claim it comes from where we log `"Unable to store offset"`? That doesn't seem right, those are different messages.

Log message have source location though, so we shouldn't have to guess.

File: library/Freckle/App/Kafka/Consumer.hs (hunk 2 above)
Comment (pbrisbin) [reply library/Freckle/App/Kafka/Consumer.hs:168]: Good question. I defer to you.

Practically speaking, I'd put my money on consumers _not_ doing smart things, but it's also trivial for them to do their own `traverse_`.

Comment (z0isch) [reply library/Freckle/App/Kafka/Consumer.hs:168]: I think `[a] -> m ()` also implies that we can't store or commit offsets on each event, so I guess for the first attempt at this I can go with the `a -> m ()` shape and see what happens.

**Comment by z0isch** [comment]
> If we wanted to keep the reduction in duplicate processing, I think still doing a store+commit on every message, but now with batching, could also help things... since you're polling less frequently it would take pressure off the broker even though you're committing more frequently. WDYT?

Probably worth a try to see what happens; however a client could still get this behavior by having `batchSize=1` right?

File: library/Freckle/App/Kafka/Consumer.hs (hunk 1 above)
Comment (z0isch) [reply library/Freckle/App/Kafka/Consumer.hs:196]: Oh the looping is from `immortal`- we call `onFinish` after every message

Comment (z0isch) [reply library/Freckle/App/Kafka/Consumer.hs:196]: Wait- that means we have a synchronous offset commit on every message. Is that what we really want?

Comment (z0isch) [reply library/Freckle/App/Kafka/Consumer.hs:196]: Opened up https://github.com/freckle/freckle-app/pull/152 to explore fixing this

**Comment by z0isch** [comment]
Closing this as the performance is back to normal with doing auto-commits for the answers consumer.

It seems like there is just a trade off on size of the batch of offset commits vs. reprocessing events. For our current consumer re-processing is fine due to the idempotency of the consumer.

It also seems like there is just going to be a performance cost to pay when deploying new consumers that is un-related to reprocessing events. The performance hit seems likely due to the kafka rebalance when adding/removing consumers to a consumer group and not due to reprocessing a handful (100s of events).

**Comment by pbrisbin** [comment]
> It also seems like there is just going to be a performance cost to pay when deploying new consumers that is un-related to reprocessing events

Did slowing the rotation from double-then-halve to plus-one-minus-one solve it?

**Comment by z0isch** [comment]
> > It also seems like there is just going to be a performance cost to pay when deploying new consumers that is un-related to reprocessing events
> 
> Did slowing the rotation from double-then-halve to plus-one-minus-one solve it?

Nope, it just made it have 3 blips instead of one. I think this makes sense because there was probably 3 re-balances
image

AFAICT it seems like the only way to avoid these performance issues is to not rebalance which I think implies doing some sort of static partition assignment for the consumers. This would require a lot of complexity in coordination between the consumers though, so I think we'd need to weight that against how important it is to not have these spikes.

**Pull request closed without merging**